### PR TITLE
Fix for the "Empty or white space strings equal 0" bug

### PR DESCRIPTION
**Pull request by griffpatch**
See: Empty or white space strings equal 0 #435

### CODE REVIEW

File: src/util/cast.js [hunk 1]
```diff
@@ -100,6 +100,12 @@ Cast.toRgbColorObject = function (value) {
 Cast.compare = function (v1, v2) {
     var n1 = Number(v1);
     var n2 = Number(v2);
+    if (n1 === 0 && (v1 == null || typeof v1 === 'string' && v1.trim().length === 0)) {
```
Comment (thisandagain) [inline src/util/cast.js]: This string of comparators should probably be moved out into it's own `util/cast` method for the sake of readability, reusability, and testing. Perhaps we could call it simply `isEmpty` and then amend the unit test suite for `compare` and add test coverage for the new method?

**Review by thisandagain — CHANGES_REQUESTED** [review]

**Comment by griffpatch** [comment]
Problem is... function calls are mega expensive when is the core runtime
code... You really want to avoid pulling things out into functions if
possible in my experience. I was very careful with my choice of comparators
here to make it as quick as possible given the key location it is being
used here? I do agree it would be nice to pull out if you could assume
javascript would re-inline it again... but in my experience it hasn't been
doing that :(

You still want me to do it?

**Comment by griffpatch** [comment]
I was wondering, I know you are currently building and running off the
non-minimized javascript for ease of debugging... Are there settings in the
build scripts to auto inline things like this (in the production build)
such that breaking it out into a function will not in the end impact the
speed of execution but instead be exactly the same? *hopeful expression*.
If this was the case it would make a big difference to how I went about
coding in these repositories?

**Comment by thisandagain** [comment]
I do. At least in the case of Chrome (v8) and Firefox (Gecko) we can be fairly certain that the JS runtime will optimize and memoize the function call for something like this. For now I'd rather err towards readability, maintenance, and test coverage than performance optimization though I do certainly appreciate that we need to find a balance between the two.

**Comment by thisandagain** [comment]
> build scripts to auto inline things like this (in the production build)
such that breaking it out into a function will not in the end impact the
speed of execution but instead be exactly the same

Good question. Not that I know of. Most webpack plugins (like UglifyJS) are more about compression than optimization. Most optimization of this type happens within each vendor's JS runtime implementation. That said, [Google's Closure Compiler](http://closure-compiler.appspot.com/home) is capable of inlining and tail call optimzation.

**Comment by griffpatch** [comment]
Just I've noticed when using the profiler in chrome that the most simple
function calls that do pretty much nothing except return values still
execute faster if they are inlined...  I'll do some benchmarking on the
change and see what impact it has before I re-commit... Might have to wait
until tomorrow.

Thanks

**Comment by thisandagain** [comment]
One of my favorite tools for testing things like this is [jsperf.com](https://jsperf.com/)

**Comment by griffpatch** [comment]
Yes, I did use a few examples on there to try things out... but I think
because of the way webpack packages the classes things are not quite as
simple as they appear when taken on their own... However, it's a great way
to test simple things :)

**Comment by thisandagain** [comment]
Yup. This is one of the reasons I've been a little hesitant to move the VM over to ES6 as adding Babel to the build process will add yet another layer of indirection.

**Comment by griffpatch** [comment]
Is Babel just a convertor from the yet to be fully implemented ES6 back to
ES5?

**Comment by thisandagain** [comment]
Yup, but it includes a series of plugins which can modify the source a fair amount and sometimes lead to indirection (particularly when dealing with performance optimization) IME.

**Comment by griffpatch** [comment]
Is the scratch-render using this already? I noted it had the 'const' and
'let' keywords and class declaration... but the debugging of those
javascript files had reverted back to ES5...

**Comment by thisandagain** [comment]
Yes. `scratch-render` uses ES6 / Babel:
https://github.com/LLK/scratch-render/blob/develop/webpack.config.js#L24

**Comment by griffpatch** [comment]
Ok, I've amended this pull request as you requested. The good news is that it doesn't appear to slow things down in the slightest! In fact it's debatably slightly faster which (very very bad timing using project #144580026). But it gives a gist which is useful in these things. Problem is the variance can be huge. Anyhow, it's no big performance issue either way so I'm quite happy with that.

**Review by thisandagain — APPROVED** [review]

**Comment by thisandagain** [comment]
Looks good. Thanks.

**Pull request merged**